File: src/platform/python/mgba/log.py
# Copyright (c) 2013-2016 Jeffrey Pfau
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
from ._pylib import ffi, lib  # pylint: disable=no-name-in-module
from . import create_callback

create_callback("mLoggerPy", "log", "_pyLog")


def install_default(logger):
    Logger.install_default(logger)


def silence():
    Logger.install_default(NullLogger())


class Logger(object):
    FATAL = lib.mLOG_FATAL
    DEBUG = lib.mLOG_DEBUG
    INFO = lib.mLOG_INFO
    WARN = lib.mLOG_WARN
    ERROR = lib.mLOG_ERROR
    STUB = lib.mLOG_STUB
    GAME_ERROR = lib.mLOG_GAME_ERROR

    _DEFAULT_LOGGER = None

    def __init__(self):
        self._handle = ffi.new_handle(self)
        self._native = ffi.gc(lib.mLoggerPythonCreate(self._handle), lib.free)

    @staticmethod
    def category_name(category):
        return ffi.string(lib.mLogCategoryName(category)).decode('UTF-8')

    @classmethod
    def install_default(cls, logger):
        cls._DEFAULT_LOGGER = logger
        lib.mLogSetDefaultLogger(logger._native)

    def log(self, category, level, message):
        print("{}: {}".format(self.category_name(category), message))


class NullLogger(Logger):
    def log(self, category, level, message):
        pass

File: src/platform/python/log.c
/* Copyright (c) 2013-2016 Jeffrey Pfau
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
#include "log.h"

static void _pyLogShim(struct mLogger* logger, int category, enum mLogLevel level, const char* format, va_list args) {
	struct mLoggerPy* pylogger = (struct mLoggerPy*) logger;
	char message[256] = {0};
	vsnprintf(message, sizeof(message) - 1, format, args);
	_pyLog(pylogger, category, level, message);
}

struct mLogger* mLoggerPythonCreate(void* pyobj) {
	struct mLoggerPy* logger = malloc(sizeof(*logger));
	logger->d.log = _pyLogShim;
	logger->d.filter = NULL;
	logger->pyobj = pyobj;
	return &logger->d;
}
